WIP: buffer: Create audio stream buffer

ktrzcinx opened the pull request:
When new component will need to process audio data in two steps,
then temporary circular buffer will be needed. Also extra information
about data stream may be needed to choose right algorithm or assert data
compatibility.

Currently only 'comp_buffer' has such a possibility but it is component
defined in topology - with much wider capabilities - and shouldn't be
created inside component just to hold temporary data stream.

Signed-off-by: Karol Trzcinski <karolx.trzcinski@linux.intel.com>

paulstelian97 commented on src/include/sof/audio/buffer.h:
```diff
@@ -58,7 +58,6 @@ struct comp_buffer {
 
 	/* runtime data */
 	uint32_t size;	/* runtime buffer size in bytes (period multiple) */
-	uint32_t alloc_size;	/* allocated size in bytes */
```
I assume the allocator itself tracks this and buffer resizes just call the allocator itself? Or do buffers not resize at all or something?

ktrzcinx replied:
This variable is only set in one placed (deleted in this commit) and has never been ridden.

dbaluta commented on src/include/sof/audio/audio_stream.h:
```diff
@@ -0,0 +1,28 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2017 Intel Corporation. All rights reserved.
+ *
```
I assume in Poland is 2020, too? Right :)

dbaluta commented on src/include/sof/audio/buffer.h:
```diff
@@ -8,6 +8,7 @@
 #ifndef __SOF_AUDIO_BUFFER_H__
 #define __SOF_AUDIO_BUFFER_H__
 
+#include <sof/audio/audio_stream.h>
```
What with the todo in the subjec?

ktrzcinx replied:
This commit is quite large and is under development

dbaluta commented on src/include/sof/audio/audio_stream.h:
```diff
@@ -0,0 +1,259 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ *
+ * Copyright(c) 2020 Intel Corporation. All rights reserved.
```
@ktrzcinx there are some inconsistencies in the style. Some functions are starting with stream some others with audio_stream.

Please use one and stick with it. I think it make sense for everthing to start with stream since we are already in audio/ directory. Including the file should be named audio/stream.h instead of audio/audio_stream.h

Also struct audio_stream should go as struct stream .

ktrzcinx replied:
Yea, I also thought for a while about name. I thought about 'stream' but there are file ipc/stream.h so  I added audio_ prefix to make it more unique, but audio_stream_function_name() lead to very long function names, so here I reject audio_prefix. 

Maybe 'astream' or 'audios' are better solutions? In terms of uniqueness and number of letters in prefix. Or maybe some acronym? For example 'cas' - circular audio stream - it will be unique and short, less self-descriptive, but it will be common structure, so maybe?

dbaluta replied:
@ktrzcinx Indeed naming is hard. We should also have some conventions like kernel has for internal/external APIs.

ktrzcinx replied:
@dbaluta by "special type of audio buffer, instead of using a generic buffer" you mean that it's not just a circular buffer but have extra fields like audio format, channels and rate ? If yes, then I can say that we are making audio firmware :P so it is our priority to have user-friendly api for audio processing. Moreover byte-stream without information about format is some kind of black box.

And what would you like to choose in naming issue? Maybe I stick to structure name 'audio_stream' but change functions prefix to 'as_'? For me it sound quite descriptive and short in typing? But I would like to here some approval of this concept or another proposition before changes.

lgirdwood replied:
my understanding now (by mentally squashing patch 2 & 3) is that this is generic buffer code.

dbaluta replied:
i dont see as_ prefix to be very useful. If there is no better option just use audio_stream even if the lines are larger.

lgirdwood replied:
@dbaluta good point - @ktrzcinx can you update.

ktrzcinx commented:
@lgirdwood I see that in currently we have tendency to duplications, see hb struct in sof/audio/kpb.h and as I far as I know there are similar issue currently under development. So I see need to split comp_buffer to part responsible directly for holding audio samples in circular buffer, base argument for data processing functions and universal brick for nesting when it is such a need.
As you can see I change almost each processing function in code and nowhere were used more fields than are present in this new structure - as usual there are need to only have circular buffer implementation with channel numbers and frame format.

lgirdwood commented:
@ktrzcinx one more question, will there be further updates to kpb to deduplicate code and fully use this update ?

ktrzcinx commented:
@lgirdwood do You want to just fully merge patch #2 i #3 ?

ktrzcinx commented:
@lgirdwood I updated prefixes to 'audio_stream_'.
About updates in kpb, my priority is to do some new features, where structure like audio_stream will be needed as internal processing buffer, but firstly I need to have this and PR merged.

dbaluta commented on src/audio/asrc/asrc.c:
```diff
@@ -64,11 +70,7 @@ struct comp_data {
-			  int *consumed,
-			  int *produced);
+	asrc_proc_func asrc_func;		/* Processing functrion */
 };
```
This should really be a separate patch. We can even merge it independently.

dbaluta commented on src/audio/detect_test.c:
```diff
@@ -86,8 +86,8 @@ struct comp_data {
-			    struct comp_buffer *source, uint32_t frames);
+	void (*detect_func)(const struct comp_dev *dev,
+			    const struct audio_stream *source, uint32_t frames);
 };
```
Adding const should to structures should be a separate patch with a proper reasoning. We can even merge it earlier, then the rest of the patches if it makes sense.

ktrzcinx replied:
It is double work, and this changes do not firmware behavior - it's only compilation time restrictions.

ktrzcinx replied:
So maybe You merge https://github.com/thesofproject/sof/pull/2273 first, to this effort little smaller.

dbaluta replied:
@ktrzcinx done!

ktrzcinx replied:
@dbaluta thanks!

dbaluta commented on src/audio/dai.c:
```diff
@@ -109,27 +110,30 @@ static void dai_dma_cb(void *arg, enum notify_id type, void *data)
 	}
 
+	sink_bytes = samples *
+		     audio_stream_sample_bytes(&dd->local_buffer->stream);
```
Simplifying code adding sink_bytes should be a separate patch.

dbaluta commented on src/audio/dai.c:
```diff
@@ -279,10 +283,10 @@ static int dai_capture_params(struct comp_dev *dev, uint32_t period_bytes,
 	uint32_t fifo;
 	int err;
+	uint32_t out_frm = dd->local_buffer->stream.frame_fmt;
 
```
Same here. Use fmt = dd->local_buffer->stream.frame_fmt;

dbaluta commented on src/audio/eq_iir/eq_iir.c:
```diff
@@ -56,20 +56,17 @@ struct comp_data {
-			    struct comp_buffer *sink,
-			    uint32_t frames);
+	eq_iir_func eq_iir_func;		/**< processing function */
 };
```
This should also be a separate patch.

dbaluta commented:
@ktrzcinx I'm sorry but there is no way in the world I would review such a huge patch. 
buffer: Create audio stream buffer should be split at least in:
  - patch adding typedefs to function
  - patch adding const to structures
  - patch introducing audio_stream

mrajwa commented:
@dbaluta, he took your advice too seriously... creating so many, unnecessary patches makes bisecting a painful procedure...

ktrzcinx commented:
SOFCI TEST

paulstelian97 commented on src/audio/asrc/asrc.c:
```diff
@@ -44,6 +44,12 @@
 	trace_error_comp(TRACE_CLASS_SRC, comp_ptr,		\
 			 __e, ##__VA_ARGS__)
 
```
Typo in commit description: "pointerts" should be "pointers".

paulstelian97 commented on src/audio/mixer.c:
```diff
@@ -77,7 +78,7 @@ static void mix_n_s16(struct comp_dev *dev, struct comp_buffer *sink,
 			dest = buffer_write_frag_s16(sink, frag);
 
-			/* Saturate to 16 bits */
+		/* Saturate to 16 bits */
```
What happened to this comment? It should be aligned with the code (I hope GH doesn't mess up the layout though)

ktrzcinx replied:
GH is write, it was some unintended change, good catch

dbaluta commented:
@mrajwa there's always needs to be a balance but let me have a look, maybe there are some patches that can be squashed together. Also, bisection is fast as soon as none of the patches breaks the build.

paulstelian97 commented on src/audio/tone.c:
```diff
@@ -143,7 +143,8 @@ static void tone_s32_default(struct comp_dev *dev, struct comp_buffer *sink,
 			}
 		}
-		tone_circ_inc_wrap(&dest, sink->end_addr, sink->size);
+		tone_circ_inc_wrap(&dest, sink->end_addr,
```
Checkpatch (line length) fix should be in a separate commit, although multiple such fixes can be brought together?

ktrzcinx replied:
It's leftover renaming, good catch

paulstelian97 commented on test/cmocka/src/audio/selector/selector_test.c:
```diff
@@ -129,7 +130,7 @@ static void verify_s16le_Xch_to_1ch(struct comp_dev *dev, struct comp_buffer *si
 	uint16_t destination;
 
-	for (i = 0; i < source->size / sizeof(uint16_t); i += in_channels) {
+	for (i = 0; i < source->size / sizeof(uint32_t); i += in_channels) {
```
I don't see why this change is not separate or the meaning of this. Can you explain?

paulstelian97 replied:
Will approve once this is clarified, everything else looks good to me.

ktrzcinx replied:
bug, changed

dbaluta commented:
@ktrzcinx I would do the following:
- create a separate PR with only the cleanups
- squash the patches that are adding const so that we have just one patch doing const
- merge this PR
- go on with adding audio stream buffer

Not sure if its github re-arranging the patches but I would have expected to see:
- first the cleanups
- cleanups should be arranged by types (e.g all the const patches should be grouped)

Anyhow, lets go and merge the cleanup patches as described above and then re-review the patch adding the audio stream buffer.

ktrzcinx commented:
> @ktrzcinx I would do the following:
> 
> create a separate PR with only the cleanups

@dbaluta you have seen this PR content 7 days ago and advised me to split it to more patches, I did as you wish, so in my opinion this advise with PR splitting is late. Also introduced cleanup is not PR-worth when there is no followed changes at horizon.
Moreover if I split this PR as you wish in last comment, there will be a lot of merge conflict (a lot of additional, pointless work) or will need to wait for first PR merge before start work on the second one (here I remind this PR is already waiting 7 days) without any profit at the end (when PR will be merged, then on the master branch it won't be any significant difference).
At least, this PR is a part of introducing new feature (this as logically independent part moved to separate PR). I split it to 3 PR (2 will come after merge), where each of them need changes from the previous one. So with current velocity it take me next 2 (or 3) week to have it done although I've got everything working locally.
It's why I'm not going to split introduced changes to another PR.

lgirdwood commented:
CI usual suspect.

dbaluta commented:
1) Adding 'const' or 'typedefing' a function is an independent change. That's trivial enough and having the trivial changes added to a cleanup PR would have got them merge much earlier leaving the reviewers a chance to concentrate on real changes.
2) Welcome to open source communities. Once you will get accustomed of how community works you will create your PR in a such manner that the trivial ones will be immediately merged while leaving us time to review to important ones.
3) I'm not here to delay your work or any things in that area. I'm here to make sure that we create code that it is easy to read and maintain. I do understand that it is frustrating and you put a lot of effort in creating this PR.
4) I have no further comments other than to look at checkpatch issues and see if there is any valid warning there.
5) You already got two approvals for this PR, someone at Intel can merge it. I'm not going to do it.

jajanusz commented:
Typo 'Moteover' in commit `fir: Mark source buffer as const in processing functions`

paulstelian97 commented:
Another typo in "audio_stream: Create wrapping function": imeplemented. I suggest you do a quick proofread of all your messages in this case.

ktrzcinx commented:
@jajanusz @paulstelian97 Thanks, I've scanned each commit message to find typos, and now should be fine

paulstelian97 commented:
Alright, IMO if CI passes this is good to merge.

jajanusz commented on src/audio/mux/mux.c:
```diff
@@ -369,7 +369,8 @@ static int mux_copy(struct comp_dev *dev)
 
 	/* produce output */
-	cd->mux(dev, sink, &sources[0], frames, &cd->config.streams[0]);
+	cd->mux(dev, sink, (const struct comp_buffer **)&sources[0],
```
Instead of casting here the declaration should have const in `struct comp_buffer *source;` variable declaration (if checkpatch does not complain)

lgirdwood commented:
@ktrzcinx @jajanusz Jenkins is usual suspects, but lets rerun internal CI as it looks like it's missing data.

lgirdwood commented:
SOFCI TEST

ktrzcinx commented:
Rebase required because of conflicts

lgirdwood commented:
@zrombel could you comment on the scan. Thanks.

zrombel commented:
Xtensa licese server issue. Rerunning build and scan. I've mode some changes to suppress those kind of issues.

lgirdwood commented:
@xiulipan checkpatch is fine, but showing an error ?

xiulipan commented:
@lgirdwood If you scroll down, you will find some warning:

```
Commit 0b45888e873f ("buffer: Create audio stream buffer")
----------------------------------------------------------
CHECK: Alignment should match open parenthesis
#2669: FILE: src/audio/src/src.c:778:
+		comp_update_buffer_consume(source,
+			consumed * audio_stream_frame_bytes(&source->stream));

CHECK: Alignment should match open parenthesis
#2675: FILE: src/audio/src/src.c:782:
+		comp_update_buffer_produce(sink,
+			produced * audio_stream_frame_bytes(&sink->stream));

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2959: FILE: src/include/sof/audio/audio_stream.h:43:
+#define audio_stream_read_frag(buffer, idx, size) \
+	audio_stream_get_frag(buffer, buffer->r_ptr, idx, size)

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2962: FILE: src/include/sof/audio/audio_stream.h:46:
+#define audio_stream_read_frag_s16(buffer, idx) \
+	audio_stream_get_frag(buffer, buffer->r_ptr, idx, sizeof(int16_t))

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2965: FILE: src/include/sof/audio/audio_stream.h:49:
+#define audio_stream_read_frag_s32(buffer, idx) \
+	audio_stream_get_frag(buffer, buffer->r_ptr, idx, sizeof(int32_t))

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2968: FILE: src/include/sof/audio/audio_stream.h:52:
+#define audio_stream_write_frag(buffer, idx, size) \
+	audio_stream_get_frag(buffer, buffer->w_ptr, idx, size)

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2971: FILE: src/include/sof/audio/audio_stream.h:55:
+#define audio_stream_write_frag_s16(buffer, idx) \
+	audio_stream_get_frag(buffer, buffer->w_ptr, idx, sizeof(int16_t))

CHECK: Macro argument reuse 'buffer' - possible side-effects?
#2974: FILE: src/include/sof/audio/audio_stream.h:58:
+#define audio_stream_write_frag_s32(buffer, idx) \
+	audio_stream_get_frag(buffer, buffer->w_ptr, idx, sizeof(int32_t))

CHECK: Alignment should match open parenthesis
#2978: FILE: src/include/sof/audio/audio_stream.h:62:
+static inline void *audio_stream_get_frag(const struct audio_stream *buffer,
+				const void *ptr, uint32_t idx, uint32_t size)

WARNING: Use safe version of memcpy - memcpy_s whenever possible
#4112: FILE: test/cmocka/src/audio/buffer/buffer_wrap.c:36:
+	memcpy(buf->stream.w_ptr, &bytes, 10);

WARNING: Use safe version of memcpy - memcpy_s whenever possible
#4125: FILE: test/cmocka/src/audio/buffer/buffer_wrap.c:45:
+	memcpy(buf->stream.w_ptr, &more_bytes, 5);

WARNING: Use safe version of memcpy - memcpy_s whenever possible
#4163: FILE: test/cmocka/src/audio/buffer/buffer_write.c:37:
+	memcpy(buf->stream.w_ptr, &bytes, 10);

WARNING: Use safe version of memcpy - memcpy_s whenever possible
#4201: FILE: test/cmocka/src/audio/buffer/buffer_write.c:72:
+	memcpy(buf->stream.w_ptr, &bytes, 10);

total: 0 errors, 4 warnings, 9 checks, 4349 lines checked

NOTE: For some of the reported defects, checkpatch may be able to
      mechanically convert to the typical style using --fix or --fix-inplace.
```

ktrzcinx commented:
@xiulipan @lgirdwood  checkpack results is located on github inside iframe with size restrictions:
`<p><iframe src="checkpatch.txt" frameborder="0" height="550" width="95%"></iframe></p>`

It's quite misleading for me too, maybe it's time to use full available size instead?

lgirdwood commented:
@xiulipan that was not visible yesterday when I checked (and I did double check as I found it odd), it was all good ? Could there be a sync issue ?

lgirdwood commented:
@ktrzcinx can you send a followup PR with the memcpy fixes as per check patch.

ktrzcinx commented:
@lgirdwood I didn't introduce memcpy functions in this place, it was before, but yes, I can fix it

paulstelian97 commented:
Memcpy functions seem to all be in unit tests? And the actual issue was old but only now discovered due to touching said code?

ktrzcinx commented:
@paulstelian97 Right

xiulipan commented:
@kakulesza @lrgirdwo @plbossart I think it would be some html issue. I wil try to and some fix to make the iframe scroll-able if the checkpatch result is very long.
